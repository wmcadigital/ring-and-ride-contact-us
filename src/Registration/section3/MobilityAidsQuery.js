import { useEffect } from "react";
import { useFormState } from "react-final-form";
import PropTypes from "prop-types";

import FormSection from "../../common/FormSection";
import ProgressIndicator from "../../common/ProgressIndicator";
import Question from "../../common/Question";
import getRequirementsSectionName from "./getRequirementsSectionName";
import CheckboxContainer from "../../common/CheckboxContainer";
import Checkbox from "../../common/Checkbox";
import TextArea from "../../common/TextArea";
import FieldError from "../../common/FieldError";
import { required } from "../../common/validation";

export const MobilityAidsOptions = (registerForYourself) => {
  if (registerForYourself) {
    return {
      manualWheelchair: "I use a manual wheelchair",
      poweredWheelchair: "I use a powered wheelchair",
      scooter: "I use a scooter",
      sticksOrCrutches: "I use a stick(s)/crutch(es)",
      walkingFrame: "I use a walking frame",
      guideDog: "I have a guide/helping dog",
      passengerLift: "I need a passenger lift",
      noMobiltyNeeds: "I don't use mobility aids",
    };
  }
  return {
    manualWheelchair: "They use a manual wheelchair",
    poweredWheelchair: "They use a powered wheelchair",
    scooter: "They use a scooter",
    sticksOrCrutches: "They use a stick(s)/crutch(es)",
    walkingFrame: "They use a walking frame",
    guideDog: "They have a guide/helping dog",
    passengerLift: "They need a passenger lift",
    noMobiltyNeeds: "They don't use mobility aids",
  };
};

const MobilityAidsQuery = ({ setMobilityAids }) => {
  const stateApi = useFormState();
  const formValues = stateApi.values;

  const registerForYourself = formValues["registerForYourself"] === "yes";

  const mobilityAidsOptions = MobilityAidsOptions(registerForYourself);

  const question = registerForYourself
    ? "Do you use a mobility aid to help you get around?"
    : `Does ${formValues["firstName"]} use a mobility aid to help themselves get around?`;

  const mobilityAidsError = stateApi.submitFailed
    ? stateApi.errors?.selectMobilityAids
    : null;

  const otherRequirementRequired = stateApi.submitFailed
    ? stateApi.errors?.mobilityAids?.otherRequirement
    : null;

  useEffect(() => {
    if (formValues.mobilityAids) {
      setMobilityAids(formValues.mobilityAids);
    }
  }, [formValues.mobilityAids]);

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition="Section 3 of 3"
        sectionName={getRequirementsSectionName(stateApi)}
      />
      <Question text={question} />
      <p>Select all that apply</p>
      <CheckboxContainer error={mobilityAidsError}>
        <FieldError text={mobilityAidsError} />
        {Object.keys(mobilityAidsOptions).map((option) => (
          <Checkbox
            key={mobilityAidsOptions[option]}
            label={mobilityAidsOptions[option]}
            fieldName={`mobilityAids.${option}`}
          />
        ))}
        <Checkbox label="Other" fieldName="mobilityAids.other" />
        {formValues.mobilityAids?.other ? (
          <TextArea
            label="Please describe your requirements"
            fieldName="mobilityAids.otherRequirement"
            validation={required}
            error={otherRequirementRequired}
          />
        ) : null}
      </CheckboxContainer>
    </FormSection>
  );
};

MobilityAidsQuery.propTypes = {
  setMobilityAids: PropTypes.func,
};

MobilityAidsQuery.defaultProps = {
  setMobilityAids: () => {},
};

export default MobilityAidsQuery;
