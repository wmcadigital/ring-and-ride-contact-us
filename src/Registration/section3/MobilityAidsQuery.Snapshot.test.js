import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import MobilityAidsQuery from "./MobilityAidsQuery";

jest.mock("react-final-form");

describe("MobilityAidsQuery", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<MobilityAidsQuery />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "no",
        firstName: "Test",
        lastName: "Testerson",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<MobilityAidsQuery />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
      submitFailed: true,
      errors: { selectMobilityAids: "Select at least 1 topic" },
    }));
    const renderer = createRenderer();

    renderer.render(<MobilityAidsQuery />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an other requirement text area if other reason selected", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
        mobilityAids: { other: true },
      },
    }));
    const renderer = createRenderer();

    renderer.render(<MobilityAidsQuery />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error - other selected, no requirement", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
        mobilityAids: { other: true },
      },
      submitFailed: true,
      errors: { mobilityAids: { otherRequirement: "Required" } },
    }));
    const renderer = createRenderer();

    renderer.render(<MobilityAidsQuery />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
