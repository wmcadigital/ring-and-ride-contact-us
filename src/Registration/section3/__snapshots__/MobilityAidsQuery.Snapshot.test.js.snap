// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobilityAidsQuery individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Applicant requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Does Test use a mobility aid to help themselves get around?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error={null}
  >
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="mobilityAids.manualWheelchair"
      label="They use a manual wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.poweredWheelchair"
      label="They use a powered wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.scooter"
      label="They use a scooter"
    />
    <Checkbox
      fieldName="mobilityAids.sticksOrCrutches"
      label="They use a stick(s)/crutch(es)"
    />
    <Checkbox
      fieldName="mobilityAids.walkingFrame"
      label="They use a walking frame"
    />
    <Checkbox
      fieldName="mobilityAids.guideDog"
      label="They have a guide/helping dog"
    />
    <Checkbox
      fieldName="mobilityAids.passengerLift"
      label="They need a passenger lift"
    />
    <Checkbox
      fieldName="mobilityAids.other"
      label="Other"
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`MobilityAidsQuery individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Do you use a mobility aid to help you get around?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error={null}
  >
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="mobilityAids.manualWheelchair"
      label="I use a manual wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.poweredWheelchair"
      label="I use a powered wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.scooter"
      label="I use a scooter"
    />
    <Checkbox
      fieldName="mobilityAids.sticksOrCrutches"
      label="I use a stick(s)/crutch(es)"
    />
    <Checkbox
      fieldName="mobilityAids.walkingFrame"
      label="I use a walking frame"
    />
    <Checkbox
      fieldName="mobilityAids.guideDog"
      label="I have a guide/helping dog"
    />
    <Checkbox
      fieldName="mobilityAids.passengerLift"
      label="I need a passenger lift"
    />
    <Checkbox
      fieldName="mobilityAids.other"
      label="Other"
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`MobilityAidsQuery renders with an error - other selected, no requirement 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Do you use a mobility aid to help you get around?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer>
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="mobilityAids.manualWheelchair"
      label="I use a manual wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.poweredWheelchair"
      label="I use a powered wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.scooter"
      label="I use a scooter"
    />
    <Checkbox
      fieldName="mobilityAids.sticksOrCrutches"
      label="I use a stick(s)/crutch(es)"
    />
    <Checkbox
      fieldName="mobilityAids.walkingFrame"
      label="I use a walking frame"
    />
    <Checkbox
      fieldName="mobilityAids.guideDog"
      label="I have a guide/helping dog"
    />
    <Checkbox
      fieldName="mobilityAids.passengerLift"
      label="I need a passenger lift"
    />
    <Checkbox
      fieldName="mobilityAids.other"
      label="Other"
    />
    <TextArea
      error="Required"
      fieldName="mobilityAids.otherRequirement"
      label="Please describe your requirements"
      validation={[Function]}
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`MobilityAidsQuery renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Do you use a mobility aid to help you get around?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error="Select at least 1 topic"
  >
    <FieldError
      text="Select at least 1 topic"
    />
    <Checkbox
      fieldName="mobilityAids.manualWheelchair"
      label="I use a manual wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.poweredWheelchair"
      label="I use a powered wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.scooter"
      label="I use a scooter"
    />
    <Checkbox
      fieldName="mobilityAids.sticksOrCrutches"
      label="I use a stick(s)/crutch(es)"
    />
    <Checkbox
      fieldName="mobilityAids.walkingFrame"
      label="I use a walking frame"
    />
    <Checkbox
      fieldName="mobilityAids.guideDog"
      label="I have a guide/helping dog"
    />
    <Checkbox
      fieldName="mobilityAids.passengerLift"
      label="I need a passenger lift"
    />
    <Checkbox
      fieldName="mobilityAids.other"
      label="Other"
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`MobilityAidsQuery renders with an other requirement text area if other reason selected 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Do you use a mobility aid to help you get around?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error={null}
  >
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="mobilityAids.manualWheelchair"
      label="I use a manual wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.poweredWheelchair"
      label="I use a powered wheelchair"
    />
    <Checkbox
      fieldName="mobilityAids.scooter"
      label="I use a scooter"
    />
    <Checkbox
      fieldName="mobilityAids.sticksOrCrutches"
      label="I use a stick(s)/crutch(es)"
    />
    <Checkbox
      fieldName="mobilityAids.walkingFrame"
      label="I use a walking frame"
    />
    <Checkbox
      fieldName="mobilityAids.guideDog"
      label="I have a guide/helping dog"
    />
    <Checkbox
      fieldName="mobilityAids.passengerLift"
      label="I need a passenger lift"
    />
    <Checkbox
      fieldName="mobilityAids.other"
      label="Other"
    />
    <TextArea
      error={null}
      fieldName="mobilityAids.otherRequirement"
      label="Please describe your requirements"
      validation={[Function]}
    />
  </CheckboxContainer>
</FormSection>
`;
